cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,237 @@
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
+        """Creates a (potentially incomplete) qubit connectivity graph of the circuit.
+
+        Iterates over the moments circuit from left to right drawing edges between logical qubits
+        that:
+            (1) have degree < 2, and
+            (2) that are involved in a 2-qubit operation in the current moment.
+        At this point the graph is forest of paths and/or simple cycles. For each simple cycle, make
+        it a path by removing the last edge that was added to it.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            The (potentially incomplete) qubit connectivity graph of the circuit.
+        """
+        circuit_graph = nx.Graph()
+        edge_order = 0
+        for op in circuit.all_operations():
+            circuit_graph.add_nodes_from(op.qubits)
+            if len(op.qubits) == 2 and all(
+                circuit_graph.degree[op.qubits[i]] < 2 for i in range(2)
+            ):
+                circuit_graph.add_edge(*op.qubits, edge_order=edge_order)
+                edge_order += 1
+        found = True
+        while found:
+            try:
+                cycle = nx.find_cycle(circuit_graph)
+                edge_to_remove = max(
+                    cycle, key=lambda x: circuit_graph.edges[x[0], x[1]]['edge_order']
+                )
+                circuit_graph.remove_edge(*edge_to_remove)
+            except nx.exception.NetworkXNoCycle:
+                found = False
+        return circuit_graph
+
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps disjoint lines of logical qubits onto lines of physical qubits.
+
+        Starting from the center physical qubit on the device, attempts to map disjoint lines of
+        logical qubits given by the circuit graph onto one long line of physical qubits on the
+        device, greedily maximizing each physical qubit's degree.
+        If this mapping cannot be completed as one long line of qubits in the circuit graph mapped
+        to qubits in the device graph, the line can be split as several line segments and then we:
+            (i)   Map first line segment.
+            (ii)  Find another high degree vertex in G near the center.
+            (iii) Map the second line segment
+            (iv)  etc.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        return self._initial_mapping(circuit.freeze())
+
+    @_compat.cached_method
```

Comment:
Caching an entire (large) circuit would be expensive. In what scenario do we believe there's an added advantage of caching? Can we just cache the circuit graph instead of caching the entire circuit? Would two different circuits with same circuit graph have the same initial mapping?